protocol/contracts/templegold/SpiceAuction.sol

```diff
@@ -0,0 +1,253 @@
+    /// @notice Name of this Spice Bazaar auction
+    string public override name;
+
+    /// @notice Last time auction was started. For zero auctions, it is the contract deploy timestamp
+    uint256 private immutable _deployTimestamp;
+
+    /// @notice Keep track of the different configurations for each auction
+    mapping(uint256 auctionId => SpiceAuctionConfig config) public auctionConfigs;
+    /// @notice Keep track of claimed amounts per auction token. This helps calculate amount of auction tokens for next epoch.
+    mapping(address token => uint256 amount) private _claimedAuctionTokens;
+    /// @notice Keep track of total allocation per auction token
+    mapping(address token => uint256 amount) private _totalAuctionTokenAllocation;
+
+    constructor(
+        address _templeGold,
+        address _spiceToken,
+        address _daoExecutor,
+        string memory _name
+    ) {
+        spiceToken = _spiceToken;
+        daoExecutor = _daoExecutor;
+        templeGold = _templeGold;
+        name = _name;
+        _deployTimestamp = block.timestamp;
+    }
+
+    /**
+     * @notice Set config for an epoch. This enables dynamic and multiple auctions especially for vested scenarios
+     * @dev Must be set before epoch auction starts
+     * @param _config Config to set
+     */
+    function setAuctionConfig(SpiceAuctionConfig calldata _config) external onlyDAOExecutor {
+        /// @dev epoch Id is only updated when auction starts. 
+        /// @dev cannot set config for past or ongoing auction
+        uint256 currentEpochIdCache = _currentEpochId;
+        if (currentEpochIdCache > 0) {
+            EpochInfo memory info = epochs[currentEpochIdCache];
+            /// Cannot set config for ongoing auction
+            if (info.startTime <= block.timestamp && block.timestamp < info.endTime) { revert InvalidConfigOperation(); }
+        }
+        if (_config.duration < MINIMUM_AUCTION_PERIOD) { revert CommonEventsAndErrors.InvalidParam(); }
+        // startCooldown can be zero
+        if (_config.waitPeriod == 0
+            || _config.minimumDistributedAuctionToken == 0) { revert CommonEventsAndErrors.ExpectedNonZero(); }
+        if (_config.recipient == address(0)) { revert CommonEventsAndErrors.InvalidAddress(); }
+
+        currentEpochIdCache += 1;
+        auctionConfigs[currentEpochIdCache] = _config;
+        emit AuctionConfigSet(currentEpochIdCache, _config);
+    }
+
+    /// @notice Remove auction config set for last epoch
+    function removeAuctionConfig() external override onlyDAOExecutor {
+        /// only delete latest epoch if auction is not started
+        uint256 id = _currentEpochId;
+        // Cannot reset an ongoing auction
+        EpochInfo storage info = epochs[id];
+        if (info.startTime <= block.timestamp && block.timestamp < info.endTime) { revert InvalidConfigOperation(); }
+        // use `epochId+1` for config of auction which has not started
+        id += 1;
```

Comment:
1/ Considering the current flow/configuration, there is no way `epochId < _currentEpochId` or `epochId > _currentEpochId` will be valid for `removeAuctionConfig`. Because you only set config and start an auction after the last auction has ended (so no need to remove auction config of old ended auction). So  has to be `epochId == _currentEpochId` (`epochId > _currentEpochId` is invalid). 

As governance controls and sets this function `setAuctionConfig()` and `removeAuctionConfig()`, it is assumed to proceed one at a time compared to setting many configs at a time.
Also, this avoids inconsistencies with setting and deleting configs for example `epochId = 10`, but deleting `epochId = 9` etc.

Reply:
With the considerations/assumptions above, this is the udpate:
```
  function removeConfig() external onlyDAOExecutor {
        /// only delete latest epoch if auction is not started
        uint256 id = _currentEpochId;
        
        // Cannot reset an ongoing auction
        EpochInfo storage info = epochs[id];
        if (info.isActive()) { revert InvalidConfigOperation(); }
        
        /// @dev could be that `auctionStart` is triggered but there's cooldown, which is not reached (so can delete epochInfo for _currentEpochId)
        // or `auctionStart` is not triggered but `auctionConfig` is set (where _currentEpochId is not updated yet)
        
        SpiceAuctionConfig storage config = auctionConfigs[id];
        if (info.startTime - lastAuctionTriggerTimestamp == config.startCooldown) {
            /// auction was started but cooldown has not passed yet
            if (config.duration == 0) { revert InvalidConfigOperation(); }
            delete auctionConfigs[id];
            delete epochs[id];
            _currentEpochId = id - 1;
            emit AuctionConfigRemoved(id, id);
        } else {
            // `auctionStart` is not triggered but `auctionConfig` is set
            id += 1;
            config = auctionConfigs[id];
            if (config.duration == 0) { revert InvalidConfigOperation(); }
            delete auctionConfigs[id];
            emit AuctionConfigRemoved(id, 0);
        }
    }
```